core/store/src/trie/mod.rs

```diff
@@ -734,6 +744,85 @@ impl Trie {
     pub fn get_trie_nodes_count(&self) -> TrieNodesCount {
         self.storage.get_trie_nodes_count()
     }
+
+    /// Given a root and a key, it returns a MerkleProof for that key.
+    /// Note that it can be a proof of membership, or a proof of non membership. This is encoded
+    /// in the boolean of the return type.
+    ///
+    /// Hence, if [get_proof](#method.get_proof) returns (true, _), it means that the key-value are found in the trie
+    /// and a proof for that is presented. Similarly, if (false, _) is returned, it means that the key-value
+    /// pair does not belong to the trie and a proof for that is also returned.
+    ///
+    /// # Errors
+    ///
+    /// If the root or any subsequent nodes are not found in the trie, `StorageError` is returned
+    pub fn get_proof(
+        &self,
+        root: &CryptoHash,
+        key: &[u8],
+    ) -> Result<(ProofState, Vec<RawTrieNodeWithSize>), StorageError> {
```

Comment:
Since we’re settled on sending raw nodes, we could change the proof here to `Vec<Arc<[u8]>>` and then we don’t need to reconstruct the nodes or encode them again.

Reply:
@mina86 I've added your feedback. Is this what you were thinking? I might have gotten lost in the discussions. Sorry for that.